rclcpp/include/rclcpp/node.hpp

```diff
@@ -481,6 +489,81 @@ class Node : public std::enable_shared_from_this<Node>
   rclcpp::node_interfaces::NodeTimeSourceInterface::SharedPtr
   get_node_time_source_interface();
 
+  /// Return the sub-namespace, if this is a sub-node, otherwise an empty string.
+  /**
+   * The returned sub-namespace is what was given to create_sub_node() or an
+   * empty string if this is an original node instance, i.e. not a sub-node.
+   *
+   * get_namespace() will return the original node namespace, and will not
+   * include the sub-namespace if one exists.
+   * To get that you need to call the get_effective_namespace() method.
+   *
+   * \sa get_namespace()
+   * \sa get_effective_namespace()
+   * \return the sub-namespace string, not including the node's original namespace
+   */
+  RCLCPP_PUBLIC
+  const std::string &
+  get_sub_namespace() const;
+
+  /// Return the effective namespace that is used when creating entities.
+  /**
+   * The returned namespace is a combination of the node namespace and the
+   * accumulated sub-namespaces, which is used as the namespace when creating
+   * entities which have relative names.
+   *
+   * \sa get_namespace()
+   * \sa get_sub_namespace()
+   * \return the sub-namespace string, not including the node's original namespace
+   */
+  RCLCPP_PUBLIC
+  const std::string &
+  get_effective_namespace() const;
```

Comment:
I felt like I had described their difference in the doc blocks, but maybe it needs clarification.

Briefly, `get_namespace()` returns the node's namespace (which is unaffected by any sub-namespaces), `get_sub_namespace()` returns the current sub-namespace, and `get_effective_namespace()` is the namespace that will effectively be used by entities (like pub/sub/client/server) when being created.

I don't expect the user to ever need to call these functions unless they're doing something complicated or their using it to construct some very specific logging functions. `get_effective_namespace()` is purely a convenience, since it's actually not used when creating entities, so I guess we could eliminate it. But the other two, return information that is originally given by the user and that the node needs to store, so I don't see a reason to not allow the user to get the information back out of the node (keep the members private but remove the accessors).

Reply:
> I felt like I had described their difference in the doc blocks, but maybe it needs clarification.
> 
> Briefly, `get_namespace()` returns the node's namespace (which is unaffected by any sub-namespaces), `get_sub_namespace()` returns the current sub-namespace, and `get_effective_namespace()` is the namespace that will effectively be used by entities (like pub/sub/client/server) when being created.
> 
> I don't expect the user to ever need to call these functions unless they're doing something complicated or their using it to construct some very specific logging functions. `get_effective_namespace()` is purely a convenience, since it's actually not used when creating entities, so I guess we could eliminate it. But the other two, return information that is originally given by the user and that the node needs to store, so I don't see a reason to not allow the user to get the information back out of the node (keep the members private but remove the accessors).

Yeah, as mentioned, I basically understood that from reading the code.  My biggest worry is that exposing these APIs to the user is going to result in a lot of answers.ros.org questions of the form: when do I call which of these APIs?  On the other hand, you seem to be saying that these probably won't need to be called except in specialized circumstances, so maybe that concern is unwarranted.  It's not a blocker for me, so I won't press the point further.

Reply:
I understand the concern, I still would error on the side of exposing information. At some level I think it’s good if this leads to questions about which to call, as that forces people to learn about the options, which I’d prefer to them calling `get_namespace()` but receiving something different from what they assumed. 

Thanks for iterating, you or anyone comes up with a better way to organize it that might be less confusing but still provides access to available data, I’m all ears.